src/features/device-state/routes/state-patch-v3.ts

```diff
@@ -0,0 +1,384 @@
+					const filter: Filter = { is_stored_at__image_location: location };
+					if (contentHash) {
+						filter.content_hash = contentHash;
+					}
+					return filter;
+				});
+				images = (await resinApiTx.get({
+					resource: 'image',
+					options: {
+						$select: ['id', 'is_stored_at__image_location'],
+						$filter:
+							imgLocationFilter.length === 1
+								? imgLocationFilter[0]
+								: imgLocationFilter,
+					},
+				})) as Array<Pick<Image, 'id' | 'is_stored_at__image_location'>>;
+				if (imageLocations.length !== images.length) {
+					throw new UnauthorizedError();
+				}
+			}
+
+			for (const [appUuid, releaseUuids] of Object.entries(appReleaseUuids)) {
+				appReleases[appUuid] =
+					releaseUuids.size === 0
+						? []
+						: ((await resinApiTx.get({
+								resource: 'release',
+								options: {
+									$select: ['id', 'commit'],
+									$filter: {
+										commit: { $in: Array.from(releaseUuids) },
+										belongs_to__application: {
+											$any: {
+												$alias: 'a',
+												$expr: {
+													a: { uuid: appUuid },
+												},
+											},
+										},
+									},
+								},
+						  })) as Array<Pick<Release, 'id' | 'commit'>>);
+				if (appReleases[appUuid].length !== releaseUuids.size) {
+					throw new UnauthorizedError();
+				}
+			}
+		});
+
+		await sbvrUtils.db.transaction(async (tx) => {
+			const resinApiTx = api.resin.clone({
+				passthrough: { req, custom, tx },
+			});
+
+			const waitPromises: Array<PromiseLike<any>> = [];
+
+			for (const uuid of uuids) {
+				const state = body[uuid];
+
+				const { apps } = state;
+
+				let deviceBody:
+					| undefined
+					| (Pick<
+							StatePatchV3Body[string],
+							typeof v3ValidPatchFields[number]
+					  > & {
+							device_name?: string;
+							is_running__release?: number | null;
+					  }) = _.pick(state, v3ValidPatchFields);
+				let metricsBody:
+					| undefined
+					| Pick<StatePatchV3Body[string], typeof metricsPatchFields[number]> =
+					_.pick(state, metricsPatchFields);
+				if (
+					Object.keys(metricsBody).length > 0 &&
+					(await shouldUpdateMetrics(uuid))
+				) {
+					// If we should force a metrics update then merge the two together and clear `metricsBody` so
+					// that we don't try to merge it again later
+					deviceBody = { ...deviceBody, ...metricsBody };
+					metricsBody = {};
+				}
+
+				if (state.name != null) {
+					deviceBody.device_name = state.name;
+				}
+
+				const device = devices.find((d) => d.uuid === uuid);
+				if (device == null) {
+					throw new UnauthorizedError();
+				}
+				const userAppUuid = device.belongs_to__application[0].uuid;
+				if (apps != null) {
+					const release = appReleases[userAppUuid].find(
+						(r) => r.commit === apps[userAppUuid].release_uuid,
+					);
+					if (release) {
+						deviceBody.is_running__release = release.id;
+					}
+				}
+
+				if (!_.isEmpty(deviceBody)) {
```

Comment:
Assuming I'm not wrong that `deviceBody` can't be `undefined`, I guess we can use the native method here, similarly to how we did before:
```suggestion
				if (Object.keys(deviceBody).length > 0) {
```